File: bcs-services/bcs-upgrader/upgrades/u1.21.202110211130/bcs_api.go
/*
 * Tencent is pleased to support the open source community by making Blueking Container Service available.
 * Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package u1_21_202110211130

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/Tencent/bk-bcs/bcs-services/bcs-upgrader/upgrader"
)

func createProject(project bcsProject, helper upgrader.UpgradeHelper) error {

	projectByte, err := json.Marshal(project)
	if err != nil {
		return err
	}
	// TODO 在新版本中，创建项目url中不用带projectID，使用 CreateProjectPath
	url := fmt.Sprintf(ProjectPath, project.ProjectID)

	_, err = helper.ClusterManagerRequest(http.MethodPost, url, projectByte)
	return err
}

func findProject(projectID string, helper upgrader.UpgradeHelper) (*bcsProject, error) {

	url := fmt.Sprintf(ProjectPath, projectID)

	replyData, err := helper.ClusterManagerRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}

	resp := new(bcsProject)
	err = json.Unmarshal(replyData, resp)
	if err != nil {
		return nil, err
	}

	return resp, nil
}

func updateProject(project bcsProject, helper upgrader.UpgradeHelper) error {
	projectJson, err := json.Marshal(project)
	if err != nil {
		return err
	}
	url := fmt.Sprintf(ProjectPath, project.ProjectID)

	_, err = helper.ClusterManagerRequest(http.MethodPut, url, projectJson)
	return err

}

func createClusters(clusters bcsReqCreateCluster, helper upgrader.UpgradeHelper) error {
	clustersJson, err := json.Marshal(clusters)
	if err != nil {
		return err
	}
	url := fmt.Sprintf(ClusterHost, clusters.ClusterID)

	_, err = helper.ClusterManagerRequest(http.MethodPost, url, clustersJson)
	return err
}

func findCluster(clustersID string, helper upgrader.UpgradeHelper) (*bcsRespFindCluster, error) {

	url := fmt.Sprintf(ClusterHost, clustersID)

	replyData, err := helper.ClusterManagerRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}

	resp := new(bcsRespFindCluster)
	err = json.Unmarshal(replyData, resp)
	if err != nil {
		return nil, err
	}

	return resp, nil
}

func updateCluster(data bcsReqUpdateCluster, helper upgrader.UpgradeHelper) error {
	url := fmt.Sprintf(ClusterHost, data.ClusterID)

	dataJson, err := json.Marshal(data)
	if err != nil {
		return err
	}

	_, err = helper.ClusterManagerRequest(http.MethodPut, url, dataJson)
	return err
}

func createNode(data reqCreateNode, helper upgrader.UpgradeHelper) error {
	nodeJson, err := json.Marshal(data)
	if err != nil {
		return err
	}
	url := fmt.Sprintf(NODEHOST, data.ClusterID)

	_, err = helper.ClusterManagerRequest(http.MethodPost, url, nodeJson)
	return err
}

func deleteNode(data reqDeleteNode, helper upgrader.UpgradeHelper) error {
	dataJson, err := json.Marshal(data)
	if err != nil {
		return err
	}
	url := fmt.Sprintf(NODEHOST, data.ClusterID)

	_, err = helper.ClusterManagerRequest(http.MethodDelete, url, dataJson)
	return err
}

func findClusterNode(clustersID string, helper upgrader.UpgradeHelper) ([]bcsNodeListData, error) {

	url := fmt.Sprintf(NODEHOST, clustersID)

	replyData, err := helper.ClusterManagerRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	resp := make([]bcsNodeListData, 0)
	err = json.Unmarshal(replyData, &resp)
	if err != nil {
		return nil, err
	}

	return resp, nil

}

File: bcs-ui/backend/components/bcs_api.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
from typing import Dict

from django.conf import settings
from requests import PreparedRequest
from requests.auth import AuthBase

from .base import BaseHttpClient, BkApiClient, ComponentAuth, update_url_parameters


class BcsApiConfig:
    """BcsApi 系统配置对象，为 Client 提供地址等信息"""

    def __init__(self, host: str):
        self.host = host

        # BCS API 系统接口地址
        self.query_cluster_id_url = f"{host}/{{env_name}}/rest/clusters/bcs/query_by_id/"
        self.get_cluster_credentials_url = f"{host}/{{env_name}}/rest/clusters/{{bcs_cluster_id}}/client_credentials"


class BcsApiAuth(AuthBase):
    """用于调用 bcs-api 系统的鉴权对象"""

    def __init__(self, access_token: str):
        self.access_token = access_token

    def __call__(self, r: PreparedRequest):
        # 从配置文件读取访问系统的通用 Token，置入请求头中
        auth_token = f'Bearer {getattr(settings, "BCS_APIGW_TOKEN", "")}'
        r.headers['Authorization'] = auth_token
        r.headers['Content-Type'] = 'application/json'

        # 在 GET 请求参数中追加 access_token
        r.url = update_url_parameters(r.url, {'access_token': self.access_token})
        return r


class BcsApiClient(BkApiClient):
    """访问 BCS API 服务的 Client 对象

    :param auth: 包含校验信息的对象

    API 方法常用请求参数说明
    ===

    :param env_name: 集群环境，比如 stag/prod
    :param project_id: 项目 ID
    :param cluster_id: 集群 ID
    :param bcs_cluster_id: 集群在 BCS 系统中的唯一 ID
    """

    def __init__(self, auth: ComponentAuth):
        self._config = BcsApiConfig(host=settings.BCS_API_PRE_URL)
        self._client = BaseHttpClient(BcsApiAuth(auth.access_token))

    def query_cluster_id(self, env_name: str, project_id: str, cluster_id: str) -> str:
        """查询集群在 BCS-Api 中的 ID

        :returns: 集群 ID 字符串
        """
        url = self._config.query_cluster_id_url.format(env_name=env_name)
        resp = self._client.request_json(
            'GET', url, params={'project_id': project_id, 'cluster_id': cluster_id}, raise_for_status=False
        )
        return resp['id']

    def get_cluster_credentials(self, env_name: str, bcs_cluster_id: str) -> Dict:
        """
        获取访问集群 apiserver 所需的鉴权信息，比如证书、user_token、server_address_path 等

        :returns: 包含集群鉴权信息的字典
        """
        url = self._config.get_cluster_credentials_url.format(env_name=env_name, bcs_cluster_id=bcs_cluster_id)
        return self._client.request_json('GET', url, raise_for_status=False)
